udp_mux.go

```diff
@@ -26,200 +28,150 @@ type UDPMuxDefault struct {
 	closedChan chan struct{}
 	closeOnce  sync.Once
 
-	// connsIPv4 and connsIPv6 are maps of all udpMuxedConn indexed by ufrag|network|candidateType
-	connsIPv4, connsIPv6 map[string]map[ipAddr]*udpMuxedConn
+	// conns are maps of all udpMuxedConn indexed by ufrag|network|candidateType
+	conns map[string]*udpMuxedConn
 
 	addressMapMu sync.RWMutex
-
-	// remote address (ip:port) -> (localip -> udpMuxedConn)
-	addressMap map[string]map[ipAddr]*udpMuxedConn
+	addressMap   map[string]*udpMuxedConn
 
 	// buffer pool to recycle buffers for net.UDPAddr encodes/decodes
 	pool *sync.Pool
 
 	mu sync.Mutex
+
+	// for UDP connection listen at unspecified address
+	localAddrsForUnspecified []net.Addr
 }
 
 const maxAddrSize = 512
 
-// UDPMuxConn is a udp PacketConn with ReadMsgUDP and File method
-// to retrieve the destination local address of the received packet
-type UDPMuxConn interface {
-	net.PacketConn
-
-	// ReadMsgUdp used to get destination address when received a udp packet
-	ReadMsgUDP(b, oob []byte) (n, oobn, flags int, addr *net.UDPAddr, err error)
-
-	// File returns a copy of the underlying os.File.
-	// It is the caller's responsibility to close f when finished.
-	// Closing c does not affect f, and closing f does not affect c.
-	File() (f *os.File, err error)
-}
-
 // UDPMuxParams are parameters for UDPMux.
 type UDPMuxParams struct {
 	Logger  logging.LeveledLogger
-	UDPConn UDPMuxConn
+	UDPConn net.PacketConn
 }
 
 // NewUDPMuxDefault creates an implementation of UDPMux
-func NewUDPMuxDefault(params UDPMuxParams) *UDPMuxDefault {
+func NewUDPMuxDefault(params UDPMuxParams) (*UDPMuxDefault, error) {
 	if params.Logger == nil {
 		params.Logger = logging.NewDefaultLoggerFactory().NewLogger("ice")
 	}
 
+	var localAddrsForUnspecified []net.Addr
+	if addr, ok := params.UDPConn.LocalAddr().(*net.UDPAddr); !ok {
+		return nil, errInvalidAddress
+	} else if ok && addr.IP.IsUnspecified() {
+		// return nil, errListenUnspecified
+		// For unspecified addresses, the correct behavior is to return errListenUnspecified, but
+		// it will break the applications that are already using unspecified UDP connection
+		// with UDPMuxDefault, so print a warn log and create a local address list for mux.
+		params.Logger.Warn("UDPMuxDefault should not listening on unspecified address, use NewMultiUDPMuxFromPort instead")
```

Comment:
For unspecified connection, print a warn log and generate addresses list, application using unspecified connection will work but should update to NewMultiUDPMuxFromPort